src/lib/coda_base/pending_coinbase.ml

```diff
@@ -0,0 +1,499 @@
+             let%map () =
+               Coinbase_stack.Stack.assert_equal stack_before stack
+             in
+             stack_after ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+
+    let%snarkydef delete_stack t stack_before stack_after =
+      let%bind addr =
+        request_witness Address.typ
+          As_prover.(map (return ()) ~f:(fun _ -> Find_index_of_oldest_stack))
+      in
+      handle
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun stack ->
+             let%map () =
+               Coinbase_stack.Stack.assert_equal stack_before stack
+             in
+             stack_after ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+  end
+
+  type t = {tree: Merkle_tree.t; pos_list: Stack_id.t list; new_pos: Stack_id.t}
+  [@@deriving sexp, bin_io]
+
+  let create_exn () =
+    let init_hash = Stack.hash Stack.empty in
+    let hash_on_level, root_hash =
+      List.fold
+        (List.init coinbase_tree_depth ~f:(fun i -> i + 1))
+        ~init:([(0, init_hash)], init_hash)
+        ~f:(fun (hashes, cur_hash) height ->
+          let merge = Hash.merge ~height:(height - 1) cur_hash cur_hash in
+          ((height, merge) :: hashes, merge) )
+    in
+    let rec create_path height path key =
+      if height < 0 then path
+      else
+        let hash =
+          Option.value_exn
+            (List.Assoc.find ~equal:Int.equal hash_on_level height)
+        in
+        create_path (height - 1)
+          ((if key mod 2 = 0 then `Left hash else `Right hash) :: path)
+          (key / 2)
+    in
+    let rec go t key =
+      if
+        Stack_id.compare key (Stack_id.of_int @@ Int.pow 2 coinbase_tree_depth)
+        > 0
+      then t
+      else
+        let path =
+          create_path (coinbase_tree_depth - 1) [] (Stack_id.to_int key)
+        in
+        go
+          (Merkle_tree.add_path t path key Stack.empty)
+          (Option.value_exn
+             ~error:(Error.of_string "Stack_id overflowed")
+             (Stack_id.incr_by_one key))
+    in
+    { tree=
+        go
+          (Merkle_tree.of_hash ~depth:coinbase_tree_depth root_hash)
+          Stack_id.zero
+    ; pos_list= []
+    ; new_pos= Stack_id.zero }
+
+  let merkle_root t = Merkle_tree.merkle_root t.tree
+
+  let empty_merkle_root () = merkle_root (create_exn ())
+
+  let with_next_index_exn t ~is_new_stack =
+    if is_new_stack then
+      let new_pos =
+        if Stack_id.equal t.new_pos (Stack_id.of_int coinbase_stacks) then
+          Stack_id.zero
+        else
+          Option.value_exn
+            ~error:(Error.of_string "Stack_id overflowed")
+            (Stack_id.incr_by_one t.new_pos)
+      in
+      {t with pos_list= t.new_pos :: t.pos_list; new_pos}
+    else t
+
+  let latest_stack_id t ~is_new_stack =
+    if is_new_stack then Some t.new_pos
+    else match t.pos_list with [] -> None | x :: _ -> Some x
+
+  let latest_stack_exn t =
```

Comment:
I think it's confusing for an `_exn` function to return an option. I would use Option.try_with if you want to return an option here.

Reply:
Changed all the `_exn` functions to return an `Or_error.t`